stdlib/ballerina-builtin/src/main/ballerina/util/natives.bal

```diff
@@ -24,19 +25,35 @@ public struct Locale {
     string countryCode;
 }
 
+@Description {value:"Represent errors related to all base64 encoders and decoders"}
+@Field {value:"message: The error message"}
+@Field {value:"cause: The error which caused the entity error"}
+public struct Base64Error {
+    string message;
+    error[] cause;
+}
+
 @Description {value:"Returns a random UUID string"}
 @Return {value:"The random string"}
 public native function uuid () returns (string);
 
-@Description {value:"Encodes a base10 encoded string to base64."}
-@Param {value:"s: string to be encoded"}
-@Return {value:"the encoded string."}
-public native function base64Encode (string s) returns (string);
+@Description {value:"Encode a given input with Base64 encoding scheme."}
+@Param {value:"contentToBeEncoded: Content that needs to be encoded can be of type string, blob or io:ByteChannel"}
+@Param {value:"charset: Charset to be used. This is used only with the string input"}
+@Return {value:"If the given input is of type string return value will be an encoded string"}
+@Return {value:"If the given input is of type blob return value will be an encoded blob"}
+@Return {value:"If the given input is of type io:ByteChannel return value will be an encoded io:ByteChannel"}
+@Return {value:"Base64Error will get return, in case of errors"}
+public native function base64Encode ((string | blob | io:ByteChannel) contentToBeEncoded, string charset="utf-8") returns (string  | blob  | io:ByteChannel | Base64Error);
 
-@Description {value:"Decodes a base64 encoded string to base10."}
-@Param {value:"s: string to be decoded"}
-@Return {value:"the decoded string."}
-public native function base64Decode (string s) returns (string);
+@Description {value:"Decode a given input with Base64 encoding scheme."}
+@Param {value:"contentToBeDecoded: Content that needs to be decoded can be of type string, blob or io:ByteChannel"}
+@Param {value:"charset: Charset to be used. This is used only with the string input"}
+@Return {value:"If the given input is of type string return value will be a decoded string"}
+@Return {value:"If the given input is of type blob return value will be a decoded blob"}
+@Return {value:"If the given input is of type io:ByteChannel return value will be a decoded io:ByteChannel"}
+@Return {value:"Base64Error will get return, in case of errors"}
+public native function base64Decode ((string | blob | io:ByteChannel) contentToBeDecoded, string charset="utf-8") returns (string  | blob  | io:ByteChannel | Base64Error);
```

Comment:
How can a base 64 encoded value be any other type than `string`? Shouldn't this be accepting a base 64 encoded string and returning a decoded value? And how do we decided what type the decoded value is? Won't we have to return a `blob`?

Reply:
At the end of the day base64 encoded values are just 24-bit groups(3 bytes) where each group is treated as four concatenated 6-bit groups. If the number of bits in the input stream is not a multiple of 24, padding is used to make up for the remaining bits. Now each of this 6 bit group has a corresponding int value (matches with the base64 alphabet). So you convert this int values to its equivalent bit form so the initial 3 bytes now have 4 bytes after the encoding.  You can get the equivalent base64 string out of the resulting encoded byte array which in this case will give you 4 characters. You can even wrap an output stream for the encoded byte data using the Base64 encoding scheme. 
When someone has encoded a file using base64 and send it across the network, these methods allows you to get a decoded byte[] or a bytechannel (in which case, it contains the decoded stream) directly which you can use it to save the file in its original form or do any processing you want.  

We make use of Java APIs internally. You can find more info as to how this algorithm has been implemented in java with all the bit shifting in Base64.java class. 

As to the the question, how do we decided what type the decoded value is returned, it’s a one to one map. If the input is a string you get an encoded string, if its a byte[] you get an encoded byte[] etc.. API doc says what each input returns. You get the encoded version of what you give. Same for decoding.